src/artifactRegistry.ts

```diff
@@ -45,6 +56,84 @@ export class ArtifactRegistryDockerRegistryClient {
     this.repositoryFields = { project, location, repository };
   }
 
+  async getGitCommitsBetweenTags({
+    prevTag,
+    nextTag,
+    dockerImageRepository,
+  }: {
+    prevTag: string;
+    nextTag: string;
+    dockerImageRepository: string;
+  }): Promise<string[]> {
+    core.info(
+      `running diff docker tags ${prevTag} ${nextTag} ${dockerImageRepository}`,
+    );
+    // Input is relatively trusted; this is largely to prevent mistaken uses
+    // like specifying a full Docker-style repository with slashes.
+    if (dockerImageRepository.includes('/')) {
+      throw Error('repository cannot contain a slash');
+    }
+
+    const dockerInfos = (
+      await this.client.listTags({
+        parent: this.client.pathTemplates.packagePathTemplate.render({
+          ...this.repositoryFields,
+          package: dockerImageRepository,
+        }),
+      })
+    )[0];
+
+    // We want to get the minimum tag for each version, since this implies those commits
+    // made a change to the docker image so are relevant to the diff.
+    const tagBoundsMap = new Map<string, { tag: string; commit: string }>();
+    for (const dockerInfo of dockerInfos) {
+      if (!dockerInfo.version || !dockerInfo.name) continue;
+
+      const tag = this.client.pathTemplates.tagPathTemplate.match(
+        dockerInfo.name,
+      ).tag;
+
+      // If it is a number we can ignore the tag
+      if (typeof tag !== 'string') continue;
+
+      // We only care about the tags between prev and next that have a git commit
+      const gitCommitMatches = tag.match(/-g([0-9a-fA-F]+)$/);
+      if (
+        ((tag >= prevTag && tag <= nextTag) ||
+          (tag <= prevTag && tag >= nextTag)) &&
+        gitCommitMatches
+      ) {
+        const minTag = tagBoundsMap.get(dockerInfo.version);
+        if (minTag && minTag.tag > tag) {
```

Comment:
I'm not sure this is the right algorithm for handling reverts (in the source repo) properly.  For example, let's say we have commit 100 currently deployed.  105 makes some change, 110 reverts it (so that the 100 image and the 110 image are the same).  This algorithm will ignore the 110 change because it's the same as the original deployed change.  The same thing happens if there was some other change 103 too (so that 110 matches 103) — the list would include 103 and 105 but not 110.

I think the right thing to do is sort the tags and process them in order and skip tags if the image is the same as the previous one you kept, but not doing a more global "skip if *any* smaller tags have the same version".